### PR TITLE
feat: add FragmentWriter and Commiter for streaming ray write

### PR DESCRIPTION
**Pull request by eddyxu**
The current `ray.data.Datasink` API does not support writing larger than memory files. We have to provide more to support streaming large files into lance.

### CODE REVIEW

File: python/python/lance/ray/sink.py [hunk 1]
```diff
@@ -15,10 +27,93 @@
...
+    df: Union[pa.Table, "pd.DataFrame", Dict], schema: Optional[pa.Schema]
+) -> pa.Table:
+    """Convert a pandas DataFrame to pyarrow Table."""
+    from ..dependencies import _PANDAS_AVAILABLE
+    from ..dependencies import pandas as pd
+
+    if isinstance(df, dict):
+        return pa.Table.from_pydict(df, schema=schema)
+    if _PANDAS_AVAILABLE and isinstance(df, pd.DataFrame):
+        return pa.Table.from_pandas(df, schema=schema)
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:43]: do we want to do drop the metadata here? `pa.Table.from_pandas` will add metadata about the original df's schema

File: python/python/lance/ray/sink.py [hunk 2]
```diff
@@ -15,10 +27,93 @@
...
+def _write_fragment(
+    stream: Iterable[pa.Table], uri: str, schema: Optional[pa.Schema] = None
+) -> Tuple[FragmentMetadata, pa.Schema]:
+    from ..dependencies import _PANDAS_AVAILABLE
+    from ..dependencies import pandas as pd
+
+    if schema is None:
+        first = next(stream)
+        if _PANDAS_AVAILABLE and isinstance(first, pd.DataFrame):
+            schema = pa.Schema.from_pandas(first)
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:56]: the type annotation only says `Iterable[pa.Table]` not `Iterable[pa.Table | pd.DataFrame]`

File: python/python/lance/ray/sink.py [hunk 3]
```diff
@@ -15,10 +27,93 @@
...
+
+
+def _write_fragment(
+    stream: Iterable[pa.Table], uri: str, schema: Optional[pa.Schema] = None
+) -> Tuple[FragmentMetadata, pa.Schema]:
+    from ..dependencies import _PANDAS_AVAILABLE
+    from ..dependencies import pandas as pd
+
+    if schema is None:
+        first = next(stream)
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:54]: nit: can we use `peekable` from `more_itertools`? https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.peekable

File: python/python/lance/ray/sink.py [hunk 4]
```diff
@@ -15,10 +27,93 @@
...
+        if _PANDAS_AVAILABLE and isinstance(first, pd.DataFrame):
+            schema = pa.Schema.from_pandas(first)
+        else:
+            schema = first.schema
+        stream = chain([first], stream)
+
+    def record_batch_converter():
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:61]: pedantic nit:
```python
map(
  functools.partial(_pd_to_arrow, schema=schema),
  stream,
)
```

File: python/python/lance/ray/sink.py [hunk 5]
```diff
@@ -15,10 +27,93 @@
...
+    ):
+        fragments = []
+        schema = None
+        for batch in write_results:
+            for fragment_str, schema_str in batch:
+                fragment = cloudpickle.loads(fragment_str)
+                fragments.append(fragment)
+                schema = cloudpickle.loads(schema_str)
+        if self.mode in set(["create", "overwrite"]):
+            op = lance.LanceOperation.Overwrite(schema, fragments)
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:110]: this is different from the standard lance behavior I think?

Comment (chebbyChefNEQ) [reply python/python/lance/ray/sink.py:110]: I think the current behavior is
```
create + already exist => Error
append | overwrite + doesn't exist => Create
```

Comment (chebbyChefNEQ) [reply python/python/lance/ray/sink.py:110]: https://github.com/lancedb/lance/blob/main/rust/lance/src/dataset.rs#L408-L415

File: python/python/lance/ray/sink.py [hunk 6]
```diff
@@ -74,59 +165,66 @@ def write(
...
+
+    def __call__(self, batch: Union[pa.Table, "pd.DataFrame"]) -> Dict[str, Any]:
+        """Write a Batch to the Lance fragment."""
+
+        transformed = self.transform(batch)
+        if not isinstance(transformed, Generator):
+            transformed = (t for t in [transformed])
+
+        fragment, schema = _write_fragment(transformed, self.base_uri, self.schema)
+        return pa.Table.from_pydict(
```
Comment (chebbyChefNEQ) [inline python/python/lance/ray/sink.py:197]: nit: why not juse return native python dict and list here?

**Review by chebbyChefNEQ — COMMENTED** [review]
mostly a bunch of nits, one question about create/append/overwrite behavior

File: python/python/lance/ray/sink.py [hunk 7]
```diff
@@ -74,59 +165,66 @@ def write(
...
+            {
+                "fragment": [cloudpickle.dumps(fragment)],
+                "schema": [cloudpickle.dumps(schema)],
```
Comment (wjones127) [inline python/python/lance/ray/sink.py:200]: Why `cloudpickle`? These both are pickle-able?

File: python/python/lance/ray/sink.py [hunk 8]
```diff
@@ -15,10 +27,93 @@
...
+
+    if isinstance(df, dict):
+        return pa.Table.from_pydict(df, schema=schema)
+    if _PANDAS_AVAILABLE and isinstance(df, pd.DataFrame):
+        return pa.Table.from_pandas(df, schema=schema)
+    return df
+
+
+def _write_fragment(
+    stream: Iterable[pa.Table], uri: str, schema: Optional[pa.Schema] = None
```
Comment (wjones127) [inline python/python/lance/ray/sink.py:48]:
```suggestion
    stream: Iterable[pa.Table | "pd.DataFrame"], uri: str, schema: Optional[pa.Schema] = None
```

File: python/python/tests/test_ray.py [hunk 9]
```diff
@@ -49,3 +54,23 @@ def test_ray_sink(tmp_path: Path):
...
+
+
+def test_ray_committer(tmp_path: Path):
+    schema = pa.schema([pa.field("id", pa.int64()), pa.field("str", pa.string())])
+
+    ds = (
+        ray.data.range(10)
+        .map(lambda x: {"id": x["id"], "str": f"str-{x['id']}"})
+        .map_batches(LanceFragmentWriter(tmp_path, schema=schema), batch_size=5)
+        .write_datasink(LanceCommitter(tmp_path))
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/python/tests/test_ray.py:66]: `Datasink::write()` requires all tables already in memory. It means that it limits each fragment size by the worker memory.

File: python/python/lance/ray/sink.py (hunk 7 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:200]: Ray treats this function returns a `pa.Table`. Pickle them to make this schema and fragment to be stored as two columns.

File: python/python/lance/ray/sink.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:43]: do you mean passing `preserve_index=False`? or explicitly providing `schema=schema`?

File: python/python/lance/ray/sink.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:54]: Dont feel that we should introduce a new deps for this line. Lets revisit this if we need more?

File: python/python/tests/test_ray.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (c21) [reply python/python/tests/test_ray.py:66]: It's a bit hacky at this moment, but we can use this as a workaround for now.

File: python/python/lance/ray/sink.py (hunk 3 above)
Comment (chebbyChefNEQ) [reply python/python/lance/ray/sink.py:54]: sounds good. Just saying `more-itertools` might already be in our transitive dep, its a fairly common pacakge

File: python/python/lance/ray/sink.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:56]: Fixed.

File: python/python/lance/ray/sink.py (hunk 8 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:48]: Fixed.

File: python/python/lance/ray/sink.py [hunk 10]
```diff
@@ -15,10 +27,93 @@
...
+        stream = chain([first], stream)
+
+    def record_batch_converter():
+        for block in stream:
+            tbl = _pd_to_arrow(block, schema)
+            yield from tbl.to_batches()
+
+    # TODO: use format v2.
+    reader = pa.RecordBatchReader.from_batches(schema, record_batch_converter())
+    fragment = LanceFragment.create(uri, reader, fragment_id=0, schema=schema)
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/python/lance/ray/sink.py:68]: Fixed.

File: python/python/lance/ray/sink.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:43]: Done

File: python/python/lance/ray/sink.py (hunk 4 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:71]: Still need to call `to_batches()` over this generator?

Comment (chebbyChefNEQ) [reply python/python/lance/ray/sink.py:71]: oh, I see.

```
chain(map(pa.Table.to_batches, map(partial(_pd_to_arrow, schema=schema), stream)))
```
?

jk, nvm as-is is fine

File: python/python/lance/ray/sink.py (hunk 7 above)
Comment (wjones127) [reply python/python/lance/ray/sink.py]: Yeah but I mean why not use just normal `pickle` from the standard library?

**Review by wjones127 — APPROVED** [review]
Seems fine, though I still think we shouldn't need cloudpickle.

**Comment by eddyxu** [comment]
> Seems fine, though I still think we shouldn't need cloudpickle.

i see. i changed it to use std `pickle`

**Review by wjones127 — APPROVED** [review]

**Pull request merged**